### PR TITLE
8341664: ReferenceClassDescImpl cache internalName

### PR DESCRIPTION
**Pull request by wenshao**
ReferenceClassDescImpl caches internalName to avoid creating a new String object for each call, which can also simplify the implementation of classfile.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341664](https://bugs.openjdk.org/browse/JDK-8341664): ReferenceClassDescImpl cache internalName (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21373/head:pull/21373` \
`$ git checkout pull/21373`

Update a local copy of the PR: \
`$ git checkout pull/21373` \
`$ git pull https://git.openjdk.org/jdk.git pull/21373/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21373`

View PR using the GUI difftool: \
`$ git pr show -t 21373`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21373.diff">https://git.openjdk.org/jdk/pull/21373.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21373#issuecomment-2397426526)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/impl/Util.java [hunk 1]
```diff
@@ -145,10 +146,10 @@ public static String toBinaryName(ClassDesc cd) {
...
 
     public static String toInternalName(ClassDesc cd) {
-        var desc = cd.descriptorString();
-        if (desc.charAt(0) == 'L')
-            return desc.substring(1, desc.length() - 1);
-        throw new IllegalArgumentException(desc);
+        if (cd instanceof ReferenceClassDescImpl rcd) {
+            return rcd.internalName();
+        }
+        return cd.descriptorString();
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/Util.java:152]:
```suggestion
        rthrow new IllegalArgumentException(desc);
```

**Review by liach — CHANGES_REQUESTED** [review]

**Comment by liach** [comment]
The caching of class entry forms in ReferenceClassDescImpl may be controversial; I will consult with other engineers.

**Comment by wenshao** [comment]
In many scenarios where classfile is used, ClassDesc is cached by statically declaring constants, such as InvokerBytecodeGenerator/StringConcatFactory. However, internalName is generated by subString every time bytecode is built, which means that the purpose of caching ClassDesc is not fully achieved.

**Review by liach — APPROVED** [review]
Good.

**Pull request closed without merging**